platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/AbstractLongList.java

```diff
@@ -0,0 +1,540 @@
+     * calling put concurrently then use a BufferedLongListWrapper. </b>
+     *
+     * @param file The file to write into, it should not exist but its parent directory should exist
+     *             and be writable.
+     * @throws IOException If there was a problem creating or writing to the file.
+     */
+    @Override
+    public void writeToFile(final Path file) throws IOException {
+        try (final FileChannel fc = FileChannel.open(file, StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
+            // write header
+            writeHeader(fc);
+            // write data
+            writeLongsData(fc);
+        }
+    }
+
+    /**
+     * Write or rewrite header in file
+     *
+     * @param fc File channel on the file to write to
+     * @throws IOException If there was a problem writing header
+     */
+    protected final void writeHeader(final FileChannel fc) throws IOException {
+        final ByteBuffer headerBuffer = ByteBuffer.allocate(currentFileHeaderSize);
+        headerBuffer.rewind();
+        headerBuffer.putInt(CURRENT_FILE_FORMAT_VERSION);
+        headerBuffer.putInt(getNumLongsPerChunk());
+        headerBuffer.putLong(maxLongs);
+        headerBuffer.putLong(minValidIndex.get());
+        headerBuffer.flip();
+        // always write at start of file
+        MerkleDbFileUtils.completelyWrite(fc, headerBuffer, 0);
+        fc.position(currentFileHeaderSize);
+    }
+
+    /**
+     * Write the long data to file, This it is expected to be in one simple block of raw longs.
+     *
+     * @param fc The file channel to write to
+     * @throws IOException if there was a problem writing longs
+     */
+    protected abstract void writeLongsData(final FileChannel fc) throws IOException;
+
+    /**
+     * Lookup a long in data
+     *
+     * @param chunkIndex the index of the chunk the long is contained in
+     * @param subIndex   The sub index of the long in that chunk
+     * @return The stored long value at given index
+     */
+    protected abstract long lookupInChunk(final long chunkIndex, final long subIndex);
+
+    /**
+     * After invocation of this method, {@link AbstractLongList#get(long)}) calls
+     * will return {@link AbstractLongList#IMPERMISSIBLE_VALUE} for indices that
+     * are before {@code newMinValidIndex}.
+     * Also, a call to this method releases memory taken by unused byte buffers.
+     * Note that newMinValidIndex is allowed to exceed the current size of the list.
+     *
+     * @param newMinValidIndex minimal valid index of the list
+     */
+    @Override
+    public final void updateMinValidIndex(final long newMinValidIndex) {
+        if (newMinValidIndex < 0) {
+            throw new IndexOutOfBoundsException("Min valid index " + newMinValidIndex + " must be non-negative");
+        }
+
+        minValidIndex.set(newMinValidIndex);
+        shrinkIfNeeded(newMinValidIndex);
```

Comment:
`updateMinValidIndex` no longer creates a memory chunk because it has undesirable side effects: `MerkleDbDataSource` may update min valid indexes in underlying instances of `LongList` without putting any values in that list. This will result in creation of multiple unused chunks.